llpc/translator/lib/SPIRV/SPIRVReader.cpp

```diff
@@ -8308,6 +8393,157 @@ llvm::Function *SPIRVToLLVM::createLibraryEntryFunc() {
+}
+
+template <> SPIRVValue *SPIRVToLLVM::getValueForScratchBoundsCheckWrapper<SPIRVStore>(SPIRVStore *toBeWrapped) const {
+  return toBeWrapped->getDst();
+}
+
+template <> SPIRVValue *SPIRVToLLVM::getValueForScratchBoundsCheckWrapper<SPIRVLoad>(SPIRVLoad *toBeWrapped) const {
+  return toBeWrapped->getSrc();
+}
+
+// This method does preparation work for adding the scratch access bounds checks for memory operations
+// like OpLoads and OpStores. We are handling private and function memory classes here.
+// The idea behind this method is that while translating a OpLoad or OpStore which loads from local scratch memory using
+// a unknown (run-time) index into an array, we parse the operands of the corresponding access chain, add a chain of
+// ANDed compares to check if the bounds requirements are met and then do a conditional jump to either the intermediate
+// block which handles the memory operation or the final block (which will skip the memory op, if bounds checks failed).
+// As mentioned, this does preparation work like inserting the bounds checks, its up to the caller caring about
+// inserting the OpLoad or OpStore. This also sets a flag that signals the main translation loop that we want to
+// continue inserting subsequent instructions at the final block instead of just going back to the parent block.
+// Currently, OpLoads will yield zero and OpStores will just omit the store, if bounds checks failed. This is up to the
+// caller. This implementation handles nested structs using arrays and 1...n-dimensional arrays. This returns if we need
+// to insert a bounds checks and pointers to the newly created basic blocks (which can be optimized away by subsequent
+// passes, if not needed).
+//
+// An example: We have the following sequence in LLVM IR:
+//
+// entry:
+// %load = load <4 x float>, <4 x float> ... localArray[runtimeIndex][runtimeIndex * 2]
+//
+// which can cause a out of bounds access.
+//
+// However, we are converting it to a sequence like this:
+//
+// entry:
+//
+// %shouldLoad1 = icmp ult i32 runtimeIndex, localArray.dim1.upperBound
+// %shouldLoad2 = icmp ult i32 runtimeIndex * 2, localArray.dim2.upperBound
+// %shouldLoad = and i1 %shouldLoad1, %shouldLoad2
+// br i1 %shouldLoad, label %loadBlock, label %finalBlock
+//
+// loadBlock:
+// %load = load <4 x float>, <4 x float> ... localArray[runtimeIndex][runtimeIndex * 2]
+// br label %finalBlock
+//
+// finalBlock:
+// %finalResult = phi ... <4 x float> [ zeroinitializer, %entry], [ %load, loadBlock ]
+// ... use %finalResult subsequently
+//
+template <class MemOpType>
+SPIRVToLLVM::ScratchBoundsCheckInfo SPIRVToLLVM::prepareScratchBoundsCheck(MemOpType *origin) {
+  constexpr ScratchBoundsCheckInfo const nullValue{};
+  if (!getPipelineOptions()->enableScratchAccessBoundsChecks) {
+    return nullValue;
+  }
+
+  SPIRVValue *toBeWrapped = getValueForScratchBoundsCheckWrapper(origin);
```

Comment:
From what I can tell, this is the only place that accesses the origin. Does it have to be a template function then? As an alternative, could you require callers to pass you this value (either directly, or have a small helper template in the header that does it), so that this long function does not have to be templated?

Reply:
I am going to change that. Originally, I wanted to use a `dynamic_cast` but unfortunately it doesn't seem to be allowed to use RTTI here.